### DISCUSSION TITLE
How to hook into doctrine data mapping process

### DISCUSSION THREAD

**Discussion by Maxi-M** [symfony/symfony#42964] [Q&A]
Hello everyone!
I have a question about the process of doctrine mapping data from db to entity.
I need to hook into the process and have an access both to entity and raw data at the same time.
The reason behind it as follows:
There is an entity with property which contains some other object. This object is not a doctrine entity, but it is instanciated depending on DB value which is mapped to another private property in entity. The logic is incapsulated in another service, so I need an access to container or DependencyInjection too.
I could make an EntityListener on postLoad and do the magic there, but in this case I would need to make the getter to private field, which I would like to avoid.
Could someone advise the solution or give a link where something similar is discussed?
Thanks

**Comment by fd6130** [comment]
Sounds like you need something like Data Transfer Object? You can manually map your entity to the dto after fetching from database.

Either using automapper or deserialize it using serializer.

https://github.com/mark-gerarts/automapper-plus-bundle

**Comment by Nek-** [comment]
Hello @Maxi-M ,

If the object is something not instantiated by doctrine and not a service, then I think the best way to do is to think it simple and do something like this:

```php
class MyEntity
{
    private $mySpecialObject;

    public function methodThatUsesSpecialObject()
    {
        $this->getMySpecialObject()->doSomethingSpecial();
    }

    public function getMySpecialObject()
    {
        if ($this->mySpecialObject) return $this->mySpecialObject;

        return $this->mySpecialObject = new MySpecialObject('fields concerned');
    }
}
```

If you really want to hook inside doctrine for the instantiation of your objects, it's a little complicated, but actually possible. It's done in [this project](https://github.com/biig-io/DomainComponent) to inject an event dispatcher.

You mostly need to override the `Instantiator` but this instantiator is generated by the `ClassMetadata` that is itself generated by the `ClassMetadataFactory`. And since you can configure the `ClassMetadataFactory`, you can inject your own `ClassMetadata` that will use your own `Instantiator`.
